clippy_lints/src/casts/cast_possible_truncation.rs

```diff
@@ -40,9 +40,13 @@ fn apply_reductions(cx: &LateContext<'_>, nbits: u64, expr: &Expr<'_>, signed: b
                     get_constant_bits(cx, right).map_or(0, |b| b.saturating_sub(1))
                 })
             },
-            BinOpKind::Rem | BinOpKind::BitAnd => get_constant_bits(cx, right)
+            BinOpKind::Rem => get_constant_bits(cx, right)
                 .unwrap_or(u64::MAX)
                 .min(apply_reductions(cx, nbits, left, signed)),
+            BinOpKind::BitAnd => get_constant_bits(cx, right)
+                .unwrap_or(u64::MAX)
+                .min(get_constant_bits(cx, left).unwrap_or(u64::MAX))
+                .min(apply_reductions(cx, nbits, left, signed)),
```

Comment:
We should apply reductions to `right` as well, e.g. in

```rust
fn x() -> u64 {
    u64::MAX
}

let _ = ((x() & 255) & 999999) as u8;
let _ = (999999 & (x() & 255)) as u8; // it would stop this one linting
```

That & the above added to the tests should be good to go, if you could squash the commits that would be great also

A bit of a nit but a commit message that doesn't require looking up the issue # would be ideal